drivers/display/mb_display_v2.c

```diff
@@ -0,0 +1,514 @@
+static const uint32_t row_mask1 = 1 << 5;
+
+/**
+ * optimized function for fast ISR, avoids port test
+ */
+static inline void mypinset(uint8_t pin)
+{
+	nrf_gpio_port_out_set(NRF_P0, 1 << pin);
+}
+
+static inline void mypinclear(uint8_t pin)
+{
+	int port = (pin & 32) != 0;
+
+	if (port == 0) {
+		nrf_gpio_port_out_clear(NRF_P0, 1 << pin);
+	} else {
+		nrf_gpio_port_out_clear(NRF_P1, 1 << (pin - 32));
+	}
+}
+
+#else
+#error "microbit led matrix only works on boards bbc_microbit and bbc_microbit_v2."
+#endif
+#endif
+
+/**
+ * NRFX Timer access to NRF hardware timer. The timer is configured
+ * as MICROBIT_DISPLAY_TIMERn. Verify activation of NRFX_TIMERn, as
+ * this may not be enforced by configuration.
+ */
+#ifdef CONFIG_MICROBIT_DISPLAY_TIMER0
+#ifdef CONFIG_NRFX_TIMER0
+#define DISPLAY_TIMER 0
+#define DISPLAY_TIMER_IRQN TIMER0_IRQn
+#else
+#error "config. inconsistent: MICROBIT_DISPLAY_TIMER0=y and NRFX_TIMER0=n "
+#endif
+#endif
+#ifdef CONFIG_MICROBIT_DISPLAY_TIMER1
+#ifdef CONFIG_NRFX_TIMER1
+#define DISPLAY_TIMER 1
+#define DISPLAY_TIMER_IRQN TIMER1_IRQn
+#else
+#error "config. inconsistent: MICROBIT_DISPLAY_TIMER1=y and NRFX_TIMER1=n "
+#endif
+#endif
+#ifdef CONFIG_MICROBIT_DISPLAY_TIMER2
+#ifdef CONFIG_NRFX_TIMER2
+#define DISPLAY_TIMER 2
+#define DISPLAY_TIMER_IRQN TIMER1_IRQn
+#else
+#error "config. inconsistent: MICROBIT_DISPLAY_TIMER2=y and NRFX_TIMER2=n "
+#endif
+#endif
+#ifdef CONFIG_MICROBIT_DISPLAY_TIMER3
+#ifdef CONFIG_NRFX_TIMER3
+#define DISPLAY_TIMER 3
+#define DISPLAY_TIMER_IRQN TIMER3_IRQn
+#else
+#error "config. inconsistent: MICROBIT_DISPLAY_TIMER3=y and NRFX_TIMER3=n "
+#endif
+#endif
+#ifdef CONFIG_MICROBIT_DISPLAY_TIMER4
+#ifdef CONFIG_NRFX_TIMER4
+#define DISPLAY_TIMER 4
+#define DISPLAY_TIMER_IRQN TIMER4_IRQn
+#else
+#error "config. inconsistent: MICROBIT_DISPLAY_TIMER4=y and NRFX_TIMER4=n "
+#endif
+#endif
+
+static nrfx_timer_t display_timer;
+
+
+
+
+/**
+ * most driver variables in a single struct
+ */
+struct myscreen_struct {
+	int current_pixel; /* counter for the refresh */
+	volatile uint32_t image_buffer;
+
+	/* animation temporal logic done by submitting work to kernel */
+	struct k_work_delayable work;
+	int delay; /* ms delay in animation steps */
+
+	bool frame_available; /* whether to continue animation */
+
+	/* animation parameters */
+	bool scrolling; /* whether to scroll or jump from frame to frame */
+	bool loop; /* whether to repeat the animation until cancelled */
+	bool text; /* if images are taken from text buffer */
+	uint16_t textlen; /* length of text to be shown */
+
+	/* Animation logic variables */
+	int16_t current_image;
+	uint16_t frame_number;
+
+	const struct mb_image *img; /* pointer to array of images to show */
+	int8_t image_number;
+
+	/* Buffer for printed strings */
+	char str_buf[CONFIG_MICROBIT_DISPLAY_STR_MAX];
+};
+
+/* unique instance of this struct */
+static struct myscreen_struct myscreen;
```

Comment:
why can't this API implement a regular device (with data, config, a init function, etc.)?